File: openvino/classification.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <openvino/openvino.hpp>


const int point_num = 1024;
const int class_num = 10;


void pc_normalize(std::vector<float>& points)
{
	float mean_x = 0, mean_y = 0, mean_z = 0;
	for (size_t i = 0; i < point_num; ++i)
	{
		mean_x += points[3 * i];
		mean_y += points[3 * i + 1];
		mean_z += points[3 * i + 2];
	}
	mean_x /= point_num;
	mean_y /= point_num;
	mean_z /= point_num;

	for (size_t i = 0; i < point_num; ++i)
	{
		points[3 * i] -= mean_x;
		points[3 * i + 1] -= mean_y;
		points[3 * i + 2] -= mean_z;
	}

	float m = 0;
	for (size_t i = 0; i < point_num; ++i)
	{
		if (sqrt(pow(points[3 * i], 2) + pow(points[3 * i + 1], 2) + pow(points[3 * i + 2], 2)) > m)
			m = sqrt(pow(points[3 * i], 2) + pow(points[3 * i + 1], 2) + pow(points[3 * i + 2], 2));
	}

	for (size_t i = 0; i < point_num; ++i)
	{
		points[3 * i] /= m;
		points[3 * i + 1] /= m;
		points[3 * i + 2] /= m;
	}
}


void classfier(std::vector<float> & points)
{
	ov::Core core;                
	//auto model = core.compile_model("cls.onnx","CPU"); 
	auto model = core.compile_model("./cls/cls_fp16.xml","CPU"); 
	auto iq = model.create_infer_request();
	auto input = iq.get_input_tensor(0);
	auto output = iq.get_output_tensor(0);
	input.set_shape({ 1, 3, point_num });
	float* input_data_host = input.data<float>();  

	for (size_t i = 0; i < 3; i++)
	{
		for (size_t j = 0; j < point_num; j++)
		{
			input_data_host[point_num * i + j] = points[3 * j + i];
		}
	}
	
	iq.infer();
	
	float* prob = output.data<float>();
	int predict_label = std::max_element(prob, prob + class_num) - prob;
	std::cout << predict_label << std::endl;
}


int main()
{
	std::vector<float> points;
	float x, y, z, nx, ny, nz;
	char ch;
	std::ifstream infile("bed_0610.txt");
	for (size_t i = 0; i < point_num; i++)
	{
		infile >> x >> ch >> y >> ch >> z >> ch >> nx >> ch >> ny >> ch >> nz;
		points.push_back(x);
		points.push_back(y);
		points.push_back(z);
	}
	infile.close();

	pc_normalize(points);

	classfier(points);

	return 0;
}

File: openvino/classification.py
import numpy as np
from openvino.inference_engine import IECore


point_num = 1024


def pc_normalize(pc):
    centroid = np.mean(pc, axis=0)
    pc = pc - centroid
    m = np.max(np.sqrt(np.sum(pc**2, axis=1)))
    pc = pc / m
    return pc


if __name__ == '__main__':
    data = np.loadtxt('./bed_0610.txt', delimiter=',')
    point_set = data[:, 0:3]
    point_set = point_set[0:point_num, :]     
    point_set[:, 0:3] = pc_normalize(point_set[:, 0:3])
    points = np.reshape(point_set, ((1, point_num, 3)))
    points = points.swapaxes(2, 1)

    ie = IECore()
    #net = ie.read_network(model="cls.onnx")
    net = ie.read_network(model="cls/cls_fp16.xml", weights="cls/cls_fp16.bin")
    exec_net = ie.load_network(network=net, device_name="CPU")
    input_name = next(iter(net.input_info))
    infer_request_handle = exec_net.start_async(request_id=0, inputs={input_name: points})
    
    if infer_request_handle.wait(-1) == 0:
        output_layer = infer_request_handle._outputs_list[1]
        outputs = infer_request_handle.output_blobs[output_layer] 
        print(outputs.buffer)
        print(np.argmax(outputs.buffer))